### DISCUSSION TITLE
MemoryLayer, MemoryProvider or IndexedMemoryProvider

### DISCUSSION THREAD

**Discussion by charlenni** [Mapsui/Mapsui#2449] [Development]
I found, that we have 3 types of memory backed layers or providers. I find this an overkill.

Two ways to go:
1. We delete the MemoryLayer and say, that all layers need a data source
2. We delete the MemoryProvider, because it isn't useful anymore.

Any suggestions?

**Comment by pauldendulk** [comment]
With the current design the `MemoryLayer` is needed in some scenarios because the `Layer` responds a little too slow. If we redesign the draw logic (for two step rendering) we should tackle this. There is also the WriteableLayer which I initially only added as a sample. There should be no need for it.

I have some bigger changes in mind which I mentioned here and there but perhaps I should write some more about it. I did mention some of it in issues. They are big changes and some are related:
- Only one type of layer. Defined in Mapsui and sealed. All layers could have mutable DataSources.
- All layers have a related DataSource, which is not a member of the layer so the layer does not need Dispose. There are different types of DataSource which require different kinds of fetching (continues fetching like in the tile layer, discrete fetching like in WMS)
- The Map should not need Dispose.
- A centralized data fetcher (in the Map or MapControl). It uses the DataSource in the layers to fetch the data. It should have a fixed number of threads. So, even if you add more layers the number of fetch threads will not increase.
- The render is responsible for performance optimizations like we use in the RasteringLayer, RasterizingTileLayer
- Centralized data projection.
- Of course the two step rendering.

I think none of it should be in v5, but perhaps we can prepare for it.

**Comment by charlenni** [comment]
Good ideas and the most are useful. The one I could not go with is the RasteringLayer and RasterizingTileLayer. They are useful, but dump, because they don't know anything about the underlying data. Because of this reason, they have to be changed often. I think we don't need them anymore if we have a two stage rendering process. What they do is to transfer the content to a version that could be painted fast.

In another discussion I ask already, what Layer is doing and why this is the default layer (by name).

**Reply by pauldendulk** [reply]
About the performance: Perhaps, but this is not the kind of thing to have an opinion about, you have to measure. There are always surprises. In a sense it is a form of caching, it could make a difference.

> Because of this reason, they have to be changed often.

What do you mean? Code changes? Or data refreshes?

**Comment by charlenni** [comment]
I talk about changes to code. We now have on with the Skia crash. Thomas tries to cache data, where he knows nothing about the structure or use case. This isn't optimal.

**Reply by pauldendulk** [reply]
If a part of the code needs a lot of rewrites this often indicates a bad design, but the bad design does not have to be in that part, it could also be in the it depends on. Mapsui is developed in iterations. It is important when adding new functionality that you rewrite the current functionality, to make it all fit together. In the current phase we should look at what we want to add, but instead of adding it we should rewrite the current design to make adding it easy.

About the crash. It may not be such a complicated issue. We have cross thread access of a resource that is not thread save. This is a bug. If we delete the rasterizer completely we would still have that same bug in the MapInfo request.
